api/src/main/java/org/apache/iceberg/io/CheckIterator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+/**
+ * An Iterator of {@link Optional} that checks a predicate for each item based on another Iterator.
+ * If predicate is true, the optional item has content, otherwise the optional is empty.
+ *
+ * @param <T> the type of objects of the dependent Iterator
+ */
+public abstract class CheckIterator<T> implements CloseableIterator<Optional<T>> {
+  private final Iterator<T> items;
+  private boolean closed;
+
+  protected CheckIterator(Iterator<T> items) {
+    this.items = items;
+    this.closed = false;
+  }
+
+  protected abstract boolean check(T item);
```

Comment:
I don't think that "check" is specific enough. You can "check" a condition and then take either action depending on the result. It doesn't imply a relationship between the condition and some behavior. I would probably come up with a better name for this implementation and the method here. Probably `accept`. Is there a reason not to pass a `Predicate` to perform the check instead of embedding it?

Reply:
Actually, the `shouldKeep` method name in the filter classes works well.